openedx/stanford/djangoapps/auth_lagunita/api.py

```diff
@@ -0,0 +1,253 @@
+            client_secret=self.client_secret,
+        )
+        # pylint: enable=line-too-long
+        log.debug("url: %s", url)
+        access_token = None
+        try:
+            response = requests.get(url)
+        except requests.exceptions.HTTPError as exception:
+            error = "requests.exceptions: {0}".format(exception)
+        else:
+            log.error(response)
+            # pylint: disable=no-member
+            if response.status_code == requests.codes.ok:
+                try:
+                    data = response.json()
+                except ValueError as error:
+                    log.error("%s", error)
+                    data = {}
+                log.info(data)
+                access_token = data.get('access_token', access_token)
+            # pylint: enable=no-member
+        log.info('access_token: %s', access_token)
+        self.access_token = access_token
+        return access_token
+
+    def lookup_lead(self, email):
+        """
+        Lookup a lead ID, based on email address
+        """
+        log.info("Lead lookup attempted: %s", email)
+        endpoint = "leads.json?filterType={filterType}&filterValues={filterValues}".format(
+            filterType='email',
+            filterValues=email,
+        )
+        result, error = self._request(endpoint, method='get')
+        if error:
+            log.error(error)
+            lead_id = None
+            log.info("Lead lookup failed: %s", email)
+        else:
+            lead_id = result['id']
+            log.info("Lead lookup succeeded: %s, lead_id=%s", email, lead_id)
+        return lead_id
+
+    def subscribe(self, user, list_id):
+        """
+        Subscribe a user to a list
+        """
+        subscription_id = None
+        log.info("Subscription attempted for user=%s", user.email)
+        lead_id = self.upsert_lead(user)
+        if lead_id:
+            data = {
+                'input': [
+                    {
+                        'id': lead_id,
+                    },
+                ],
+            }
+            endpoint = "lists/{list_id}/leads.json".format(
+                list_id=list_id,
+            )
+            log.debug("lead_id=%s", lead_id)
+            result, error = self._request(endpoint, data)
+            log.debug(result)
+            if error:
+                log.error(error)
+                subscription_id = None
+            else:
+                subscription_id = result['id']
+        if subscription_id:
+            log.info("Subscription succeeded for user=%s, subscription_id=%s", user.email, subscription_id)
+        else:
+            log.error("Subscription failed for user=%s", user.email)
+        return subscription_id
+
+    def _create_lead(self, user):
+        """
+        Create a new lead entry, fails if already exists
+        """
+        log.info("Lead creation attempted: %s", user.email)
+        data = {
+            'action': 'createOnly',
```

Comment:
The docs show two endpoints ([first](https://developers.marketo.com/rest-api/lead-database/leads/#create_and_update), [second](https://developers.marketo.com/rest-api/lead-database/leads/#push_lead_to_marketo)) that support adding new leads, but it's not clear to me which one we want here...the code here is hitting the first one, but it doesn't support the `'reason'` and `'source'` params, while second one does, but that one additionally requires a `'programName'` param. Thoughts?